sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -120,17 +120,19 @@ object RowEncoder {
 
     case StructType(fields) =>
       val convertedFields = fields.zipWithIndex.map { case (f, i) =>
-        val method = if (f.dataType.isInstanceOf[StructType]) {
-          "getStruct"
+        val x = extractorsFor(
+          GetExternalRowField(inputObject, i, externalDataTypeForInput(f.dataType), f.nullable),
+          f.dataType
+        )
+        if (f.nullable) {
+          If(
```

Comment:
if we do the null check inside GetExternalRowField then the code for serializerFor also needs to be pushed into it (to be inside the null check), and i can not figure out how to do that...